jetty-hazelcast/src/main/java/org/eclipse/jetty/hazelcast/session/HazelcastSessionDataStore.java

```diff
@@ -0,0 +1,205 @@
+//
+//      The Apache License v2.0 is available at
+//      http://www.opensource.org/licenses/apache2.0.php
+//
+//  You may elect to redistribute this code under either of these licenses.
+//  ========================================================================
+//
+
+package org.eclipse.jetty.hazelcast.session;
+
+import com.hazelcast.core.IMap;
+import org.eclipse.jetty.server.session.AbstractSessionDataStore;
+import org.eclipse.jetty.server.session.SessionContext;
+import org.eclipse.jetty.server.session.SessionData;
+import org.eclipse.jetty.server.session.SessionDataStore;
+import org.eclipse.jetty.util.annotation.ManagedObject;
+import org.eclipse.jetty.util.log.Log;
+import org.eclipse.jetty.util.log.Logger;
+
+import java.util.Collections;
+import java.util.Set;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.stream.Collectors;
+
+/**
+ * Session data stored in Hazelcast
+ */
+@ManagedObject
+public class HazelcastSessionDataStore
+    extends AbstractSessionDataStore
+    implements SessionDataStore
+{
+
+    private  final static Logger LOG = Log.getLogger( "org.eclipse.jetty.server.session");
+
+    private IMap<String, SessionData> sessionDataMap;
+
+    public HazelcastSessionDataStore()
+    {
+        // no op
+    }
+
+    @Override
+    public SessionData load( String id )
+        throws Exception
+    {
+
+        final AtomicReference<SessionData> reference = new AtomicReference<SessionData>();
+        final AtomicReference<Exception> exception = new AtomicReference<Exception>();
+
+        //ensure the load runs in the context classloader scope
+        _context.run( () -> {
+            try
+            {
+                if (LOG.isDebugEnabled())
+                {
+                    LOG.debug( "Loading session {} from hazelcast", id );
+                }
+                SessionData sd = sessionDataMap.get( getCacheKey( id ) );
+                reference.set(sd);
+            }
+            catch (Exception e)
+            {
+                exception.set(e);
+            }
+        } );
+
+        if (exception.get() != null)
+        {
+            throw exception.get();
+        }
+        return reference.get();
+    }
+
+    @Override
+    public boolean delete( String id )
+        throws Exception
+    {
+        return sessionDataMap == null ? false : sessionDataMap.remove( getCacheKey( id ) ) != null;
+    }
+
+    public IMap<String, SessionData> getSessionDataMap()
+    {
+        return sessionDataMap;
+    }
+
+    public void setSessionDataMap( IMap<String, SessionData> sessionDataMap )
+    {
+        this.sessionDataMap = sessionDataMap;
+    }
+
+    @Override
+    public void initialize( SessionContext context )
+        throws Exception
+    {
+        _context = context;
+    }
+
+    @Override
+    public void doStore( String id, SessionData data, long lastSaveTime )
+        throws Exception
+    {
+        this.sessionDataMap.put( getCacheKey( id ), data);
```

Comment:
The other approach would be to update only dirty attributes via Hazelcast Entry Processor. I see you have a session wide [dirty flag](https://github.com/eclipse/jetty.project/blob/jetty-9.4.x/jetty-server/src/main/java/org/eclipse/jetty/server/session/SessionData.java#L57) but not attribute based one. It looks like someone started to implement it but [not finished yet](https://github.com/eclipse/jetty.project/blob/jetty-9.4.x/jetty-server/src/main/java/org/eclipse/jetty/server/session/SessionData.java#L163) assuming the argument `name` is attribute name.

You can see [Entry Processor based approach](https://github.com/hazelcast/hazelcast-wm/blob/v3.8.1/src/main/java/com/hazelcast/web/entryprocessor/SessionUpdateEntryProcessor.java#L78) in our ServletFilter based Session Replication Module. We don't have this feature in Hazelcast Tomcat/Jetty Native Implementations.